python/ambassador/config/resourceprocessor.py

```diff
@@ -0,0 +1,671 @@
+        # TODO(impl): Remove this?
+        # if namespace and not obj.get('namespace', None):
+        #     obj['namespace'] = namespace
+
+        # Brutal hackery.
+        if obj['kind'] == 'Service':
+            self.logger.debug("%s PROCESS saving service %s" % (self.location, obj['name']))
+            self.services[obj['name']] = obj
+        else:
+            # Fine. Fine fine fine.
+            serialization = dump_yaml(obj, default_flow_style=False)
+
+            try:
+                r = ACResource.from_dict(rkey, rkey, serialization, obj)
+                self.elements.append(r)
+            except Exception as e:
+                self.aconf.post_error(e.args[0])
+
+            self.logger.debug("%s PROCESS %s save %s: %s" % (self.location, obj['kind'], rkey, serialization))
+
+        return True
+
+    def emit(self, emission: ResourceEmission):
+        if self._emit(emission):
+            self.locations.current.ocount += 1
+
+
+class KubernetesProcessor:
+    """
+    An abstract processor for Kubernetes objects that emit configuration
+    resources.
+    """
+
+    def kinds(self) -> FrozenSet[KubernetesGVK]:
+        # Override kinds to describe the types of resources this processor wants
+        # to process.
+        return frozenset()
+
+    def _process(self, obj: KubernetesObject) -> None:
+        # Override _process to handle a single resource. Note that the entry
+        # point for _process is try_process; _process should not be called
+        # directly.
+        pass
+
+    def try_process(self, obj: KubernetesObject) -> bool:
+        if obj.gvk not in self.kinds():
+            return False
+
+        self._process(obj)
+        return True
+
+    def finalize(self) -> None:
+        # Override finalize to do processing at the end of the configuration
+        # fetching.
+        pass
+
+
+class ManagedKubernetesProcessor (KubernetesProcessor):
+    """
+    An abstract processor that provides access to a resource manager.
+    """
+
+    manager: ResourceManager
+
+    def __init__(self, manager: ResourceManager):
+        self.manager = manager
+
+    @property
+    def aconf(self) -> Config:
+        return self.manager.aconf
+
+    @property
+    def logger(self) -> logging.Logger:
+        return self.manager.logger
+
+
+class AmbassadorProcessor (ManagedKubernetesProcessor):
+    """
+    A Kubernetes object processor that emits direct IR from an Ambassador CRD.
+    """
+
+    def kinds(self) -> FrozenSet[KubernetesGVK]:
+        kinds = [
+            'AuthService',
+            'ConsulResolver',
+            'Host',
+            'KubernetesEndpointResolver',
+            'KubernetesServiceResolver',
+            'LogService',
+            'Mapping',
+            'Module',
+            'RateLimitService',
+            'TCPMapping',
+            'TLSContext',
+            'TracingService',
+        ]
+
+        return frozenset([KubernetesGVK.for_ambassador(kind) for kind in kinds])
```

Comment:
Right now this will only handle Kubernetes objects that have v2 as their Ambassador API version. Will we need to also support v0 and v1? (Note this only applies to actual Kubernetes objects, not annotations.)